File: UDF Dönüştürücü/Core/Tools/UDFStructureParser.swift
import Foundation

/// UYAP content.xml içindeki <elements> bölümünü yapılandırılmış düzenleme modeline çevirir.
enum UDFStructureParser {

    static func parse(rawXML: String, plainText: String) -> UDFEditDocument {
        let headers = parseHeaderFooterContainer(rawXML, container: "headers", item: "header", plainText: plainText)
        let footers = parseHeaderFooterContainer(rawXML, container: "footers", item: "footer", plainText: plainText)

        guard rawXML.contains("<elements") else {
            var doc = fallbackDocument(from: plainText)
            doc.headers = headers
            doc.footers = footers
            return doc
        }

        guard let elementsBody = extractElementsBody(from: rawXML) else {
            var doc = fallbackDocument(from: plainText)
            doc.headers = headers
            doc.footers = footers
            return doc
        }

        var blocks: [UDFEditBlock] = []
        var search = elementsBody.startIndex

        while search < elementsBody.endIndex {
            if let range = elementsBody.range(of: "<paragraph ", range: search..<elementsBody.endIndex),
               let end = elementsBody.range(of: "</paragraph>", range: range.upperBound..<elementsBody.endIndex) {
                let xml = String(elementsBody[range.lowerBound..<end.upperBound])
                blocks.append(.paragraph(parseParagraph(xml, plainText: plainText)))
                search = end.upperBound
                continue
            }

            if let range = elementsBody.range(of: "<table", range: search..<elementsBody.endIndex),
               let end = elementsBody.range(of: "</table>", range: range.upperBound..<elementsBody.endIndex) {
                let xml = String(elementsBody[range.lowerBound..<end.upperBound])
                blocks.append(.table(parseTable(xml, plainText: plainText)))
                search = end.upperBound
                continue
            }

            if let nextPara = elementsBody.range(of: "<paragraph ", range: search..<elementsBody.endIndex),
               let nextTable = elementsBody.range(of: "<table", range: search..<elementsBody.endIndex) {
                search = min(nextPara.lowerBound, nextTable.lowerBound)
            } else if let next = elementsBody.range(of: "<paragraph ", range: search..<elementsBody.endIndex)
                ?? elementsBody.range(of: "<table", range: search..<elementsBody.endIndex) {
                search = next.lowerBound
            } else {
                break
            }
        }

        if blocks.isEmpty {
            var doc = fallbackDocument(from: plainText)
            doc.headers = headers
            doc.footers = footers
            return doc
        }
        return UDFEditDocument(headers: headers, blocks: blocks, footers: footers)
    }

    // MARK: - Header / Footer

    private static func parseHeaderFooterContainer(
        _ xml: String,
        container: String,
        item: String,
        plainText: String
    ) -> [UDFEditHeaderFooter] {
        guard let body = extractTaggedBody(xml, openTag: "<\(container)", closeTag: "</\(container)>") else {
            return []
        }

        var items: [UDFEditHeaderFooter] = []
        var search = body.startIndex
        let openItem = "<\(item)"

        while let start = body.range(of: openItem, range: search..<body.endIndex),
              let end = body.range(of: "</\(item)>", range: start.upperBound..<body.endIndex) {
            let itemXML = String(body[start.lowerBound..<end.upperBound])
            let type = extractStringAttr(itemXML, name: "type") ?? "default"
            var paragraphs: [UDFEditParagraph] = []
            var pSearch = itemXML.startIndex

            while let pStart = itemXML.range(of: "<paragraph ", range: pSearch..<itemXML.endIndex),
                  let pEnd = itemXML.range(of: "</paragraph>", range: pStart.upperBound..<itemXML.endIndex) {
                let pXML = String(itemXML[pStart.lowerBound..<pEnd.upperBound])
                paragraphs.append(parseParagraph(pXML, plainText: plainText))
                pSearch = pEnd.upperBound
            }

            if paragraphs.isEmpty,
               let pStart = itemXML.range(of: "<paragraph", range: itemXML.startIndex..<itemXML.endIndex),
               let pEnd = itemXML.range(of: "</paragraph>", range: pStart.upperBound..<itemXML.endIndex) {
                let pXML = String(itemXML[pStart.lowerBound..<pEnd.upperBound])
                paragraphs.append(parseParagraph(pXML, plainText: plainText))
            }

            if paragraphs.isEmpty {
                paragraphs = [UDFEditParagraph(runs: [UDFEditRun(text: "")])]
            }

            items.append(UDFEditHeaderFooter(type: type, paragraphs: paragraphs))
            search = end.upperBound
        }

        return items
    }

    // MARK: - Paragraph

    private static func parseParagraph(_ xml: String, plainText: String) -> UDFEditParagraph {
        UDFEditParagraph(
            alignment: extractIntAttr(xml, name: "Alignment") ?? 3,
            spaceAbove: extractFloatAttr(xml, name: "SpaceAbove") ?? 1,
            spaceBelow: extractFloatAttr(xml, name: "SpaceBelow") ?? 1,
            leftIndent: extractFloatAttr(xml, name: "LeftIndent") ?? 0,
            rightIndent: extractFloatAttr(xml, name: "RightIndent") ?? 0,
            firstLineIndent: extractFloatAttr(xml, name: "FirstLineIndent") ?? 0,
            hangingIndent: extractFloatAttr(xml, name: "Hanging") ?? 0,
            lineSpacing: extractFloatAttr(xml, name: "LineSpacing") ?? 0,
            tabStops: parseTabStops(xml),
            runs: parseRuns(in: xml, plainText: plainText)
        )
    }

    private static func parseRuns(in xml: String, plainText: String) -> [UDFEditRun] {
        let chars = Array(plainText)
        struct TagRun {
            let kind: UDFEditRun.Kind
            let startOffset: Int
            let length: Int
            let bold: Bool
            let italic: Bool
            let underline: Bool
            let fontSize: CGFloat?
            let fontFamily: String?
            let foreground: Int?
            let background: Int?
        }

        var tagged: [TagRun] = []
        let patterns: [(String, UDFEditRun.Kind)] = [
            ("<content ", .content),
            ("<field ", .field(name: "")),
            ("<space ", .space)
        ]

        for (pattern, kind) in patterns {
            var search = xml.startIndex
            while let tagStart = xml.range(of: pattern, range: search..<xml.endIndex),
                  let tagEnd = xml.range(of: "/>", range: tagStart.upperBound..<xml.endIndex) {
                let tag = String(xml[tagStart.lowerBound..<tagEnd.upperBound])
                guard let start = extractIntAttr(tag, name: "startOffset"),
                      let length = extractIntAttr(tag, name: "length") else {
                    search = tagEnd.upperBound
                    continue
                }

                let resolvedKind: UDFEditRun.Kind
                if case .field = kind {
                    let name = extractStringAttr(tag, name: "fieldName")
                        ?? extractStringAttr(tag, name: "name")
                        ?? extractStringAttr(tag, name: "FieldName")
                        ?? "alan"
                    resolvedKind = .field(name: name)
                } else {
                    resolvedKind = kind
                }

                tagged.append(TagRun(
                    kind: resolvedKind,
                    startOffset: start,
                    length: length,
                    bold: tag.localizedCaseInsensitiveContains("bold=\"true\""),
                    italic: tag.localizedCaseInsensitiveContains("italic=\"true\""),
                    underline: tag.localizedCaseInsensitiveContains("underline=\"true\""),
                    fontSize: extractFloatAttr(tag, name: "size") ?? extractFloatAttr(tag, name: "Size"),
                    fontFamily: extractStringAttr(tag, name: "family") ?? extractStringAttr(tag, name: "Family"),
                    foreground: UDFColorCodec.parse(extractStringAttr(tag, name: "foreground"))
                        ?? UDFColorCodec.parse(extractStringAttr(tag, name: "Foreground")),
                    background: UDFColorCodec.parse(extractStringAttr(tag, name: "background"))
                        ?? UDFColorCodec.parse(extractStringAttr(tag, name: "Background"))
                ))
                search = tagEnd.upperBound
            }
        }

        tagged.sort { $0.startOffset < $1.startOffset }

        return tagged.map { tag in
            let text = sliceText(from: chars, start: tag.startOffset, length: tag.length)
            return UDFEditRun(
                kind: tag.kind,
                text: text,
                isBold: tag.bold,
                isItalic: tag.italic,
                isUnderline: tag.underline,
                fontSize: tag.fontSize ?? 12,
                fontFamily: tag.fontFamily ?? "Times New Roman",
                foregroundARGB: tag.foreground,
                backgroundARGB: tag.background
            )
        }
    }

    // MARK: - Table

    private static func parseTable(_ xml: String, plainText: String) -> UDFEditTable {
        let columnCount = extractIntAttr(xml, name: "columnCount") ?? 2
        let spansString = extractStringAttr(xml, name: "columnSpans") ?? ""
        let columnSpans = spansString.split(separator: ",").compactMap { Int($0.trimmingCharacters(in: .whitespaces)) }
        let border = extractStringAttr(xml, name: "border") ?? "borderCell"

        var rows: [UDFEditTableRow] = []
        var search = xml.startIndex
        while let rowStart = xml.range(of: "<row", range: search..<xml.endIndex),
              let rowEnd = xml.range(of: "</row>", range: rowStart.upperBound..<xml.endIndex) {
            let rowXML = String(xml[rowStart.lowerBound..<rowEnd.upperBound])
            rows.append(parseTableRow(rowXML, plainText: plainText))
            search = rowEnd.upperBound
        }

        return UDFEditTable(
            columnCount: max(columnCount, 1),
            columnSpans: columnSpans,
            border: border,
            rows: rows
        )
    }

    private static func parseTableRow(_ xml: String, plainText: String) -> UDFEditTableRow {
        let rowType = extractStringAttr(xml, name: "rowType") ?? "dataRow"
        var cells: [UDFEditTableCell] = []
        var search = xml.startIndex

        while let cellStart = xml.range(of: "<cell", range: search..<xml.endIndex),
              let cellEnd = xml.range(of: "</cell>", range: cellStart.upperBound..<xml.endIndex) {
            let cellXML = String(xml[cellStart.lowerBound..<cellEnd.upperBound])
            cells.append(parseTableCell(cellXML, plainText: plainText))
            search = cellEnd.upperBound
        }

        return UDFEditTableRow(rowType: rowType, cells: cells)
    }

    private static func parseTableCell(_ xml: String, plainText: String) -> UDFEditTableCell {
        let colspan = extractIntAttr(xml, name: "colspan") ?? 1
        let rowspan = extractIntAttr(xml, name: "rowspan") ?? 1
        let fill = UDFColorCodec.parse(extractStringAttr(xml, name: "fillColor"))
            ?? UDFColorCodec.parse(extractStringAttr(xml, name: "background"))

        var paragraphs: [UDFEditParagraph] = []
        var search = xml.startIndex
        while let pStart = xml.range(of: "<paragraph ", range: search..<xml.endIndex),
              let pEnd = xml.range(of: "</paragraph>", range: pStart.upperBound..<xml.endIndex) {
            let pXML = String(xml[pStart.lowerBound..<pEnd.upperBound])
            paragraphs.append(parseParagraph(pXML, plainText: plainText))
            search = pEnd.upperBound
        }

        if paragraphs.isEmpty,
           let pStart = xml.range(of: "<paragraph", range: xml.startIndex..<xml.endIndex),
           let pEnd = xml.range(of: "</paragraph>", range: pStart.upperBound..<xml.endIndex) {
            let pXML = String(xml[pStart.lowerBound..<pEnd.upperBound])
            paragraphs.append(parseParagraph(pXML, plainText: plainText))
        }

        if paragraphs.isEmpty {
            paragraphs = [UDFEditParagraph(runs: [UDFEditRun(text: "")])]
        }

        return UDFEditTableCell(
            colspan: max(colspan, 1),
            rowspan: max(rowspan, 1),
            fillColorARGB: fill,
            paragraphs: paragraphs
        )
    }

    // MARK: - Helpers

    private static func fallbackDocument(from text: String) -> UDFEditDocument {
        let paragraphs = text.components(separatedBy: "\n").map { line -> UDFEditParagraph in
            UDFEditParagraph(alignment: 3, runs: [UDFEditRun(text: line)])
        }
        return UDFEditDocument(blocks: paragraphs.map { UDFEditBlock.paragraph($0) })
    }

    private static func extractTaggedBody(_ xml: String, openTag: String, closeTag: String) -> String? {
        guard let start = xml.range(of: openTag) else { return nil }
        guard let openEnd = xml.range(of: ">", range: start.upperBound..<xml.endIndex) else { return nil }
        guard let close = xml.range(of: closeTag, range: openEnd.upperBound..<xml.endIndex) else { return nil }
        return String(xml[openEnd.upperBound..<close.lowerBound])
    }

    private static func extractElementsBody(from xml: String) -> String? {
        guard let start = xml.range(of: "<elements") else { return nil }
        guard let openEnd = xml.range(of: ">", range: start.upperBound..<xml.endIndex) else { return nil }
        guard let close = xml.range(of: "</elements>", range: openEnd.upperBound..<xml.endIndex) else { return nil }
        return String(xml[openEnd.upperBound..<close.lowerBound])
    }

    private static func sliceText(from chars: [Character], start: Int, length: Int) -> String {
        guard start >= 0, length >= 0, start < chars.count else { return "" }
        let end = min(start + length, chars.count)
        guard start < end else { return "" }
        return String(chars[start..<end])
    }

    private static func parseTabStops(_ xml: String) -> [CGFloat] {
        guard let tabStr = extractStringAttr(xml, name: "TabSet") else { return [] }
        let parts = tabStr.split(separator: ":")
        if let pos = parts.first.flatMap({ Double($0) }).map({ CGFloat($0) }), pos > 0 {
            return [pos]
        }
        return []
    }

    private static func extractIntAttr(_ xml: String, name: String) -> Int? {
        for key in [name, name.lowercased(), name.capitalized] {
            guard let range = xml.range(of: "\(key)=\"") else { continue }
            let start = range.upperBound
            guard let end = xml.range(of: "\"", range: start..<xml.endIndex) else { continue }
            if let val = Int(xml[start..<end.lowerBound]) { return val }
        }
        return nil
    }

    private static func extractFloatAttr(_ xml: String, name: String) -> CGFloat? {
        for key in [name, name.lowercased(), name.capitalized] {
            guard let range = xml.range(of: "\(key)=\"") else { continue }
            let start = range.upperBound
            guard let end = xml.range(of: "\"", range: start..<xml.endIndex) else { continue }
            if let val = Double(xml[start..<end.lowerBound]) { return CGFloat(val) }
        }
        return nil
    }

    private static func extractStringAttr(_ xml: String, name: String) -> String? {
        for key in [name, name.lowercased(), name.capitalized] {
            guard let range = xml.range(of: "\(key)=\"") else { continue }
            let start = range.upperBound
            guard let end = xml.range(of: "\"", range: start..<xml.endIndex) else { continue }
            let val = String(xml[start..<end.lowerBound])
            if !val.isEmpty { return val }
        }
        return nil
    }
}

File: app/src/main/java/com/velikececi/udfdonusturucu/core/tools/editor/UdfStructureParser.kt
package com.velikececi.udfdonusturucu.core.tools.editor

/**
 * UDFStructureParser.swift'in Kotlin karşılığı. UYAP content.xml içindeki `<elements>` bölümünü
 * yapılandırılmış düzenleme modeline çevirir.
 *
 * Swift'in `Character` tabanlı String aralıklarının yerine Kotlin'in UTF-16 kod birimi tabanlı
 * `indexOf`/`substring`'i kullanılır — Türkçe alfabesi (ç ğ ı İ ö ş ü dahil) tamamen Temel Çok
 * Dilli Düzlem (BMP) içinde tek kod birimli olduğundan bu iki yaklaşım pratikte eşdeğerdir.
 * [UdfStructureWriter] de aynı UTF-16 tabanlı ofsetlerle yazdığından, Android'in kendi
 * yaz→oku döngüsü her koşulda kendi içinde tutarlıdır.
 */
object UdfStructureParser {

    fun parse(rawXml: String, plainText: String): UdfEditDocument {
        val headers = parseHeaderFooterContainer(rawXml, container = "headers", item = "header", plainText = plainText)
        val footers = parseHeaderFooterContainer(rawXml, container = "footers", item = "footer", plainText = plainText)

        if (!rawXml.contains("<elements")) {
            return fallbackDocument(plainText).copy(headers = headers, footers = footers)
        }

        val elementsBody = extractElementsBody(rawXml)
            ?: return fallbackDocument(plainText).copy(headers = headers, footers = footers)

        val blocks = mutableListOf<UdfEditBlock>()
        var search = 0

        // iOS kaynağından KASITLI SAPMA: Swift sürümü her turda önce "<paragraph "i (aramanın
        // KALAN tamamında, en yakın olan değil) dener; bu, bir <table> ilk paragraftan önce
        // geldiğinde o tabloyu sessizce atlayıp veri kaybına yol açıyordu. Burada iki etiketten
        // METİNDE DAHA ÖNCE geleni işlenir — UYAP'tan gelen gerçek dosyalarda (ör. üstte tablo
        // içeren dilekçeler) veri kaybını önler.
        while (search < elementsBody.length) {
            val paraStart = elementsBody.indexOf("<paragraph ", search)
            val tableStart = elementsBody.indexOf("<table", search)

            val nextIsParagraph = when {
                paraStart == -1 -> false
                tableStart == -1 -> true
                else -> paraStart <= tableStart
            }
            if (paraStart == -1 && tableStart == -1) break

            if (nextIsParagraph) {
                val end = elementsBody.indexOf("</paragraph>", paraStart)
                if (end == -1) break
                val xml = elementsBody.substring(paraStart, end + "</paragraph>".length)
                blocks.add(UdfEditBlock.ParagraphBlock(parseParagraph(xml, plainText)))
                search = end + "</paragraph>".length
            } else {
                val end = elementsBody.indexOf("</table>", tableStart)
                if (end == -1) break
                val xml = elementsBody.substring(tableStart, end + "</table>".length)
                blocks.add(UdfEditBlock.TableBlock(parseTable(xml, plainText)))
                search = end + "</table>".length
            }
        }

        if (blocks.isEmpty()) {
            return fallbackDocument(plainText).copy(headers = headers, footers = footers)
        }
        return UdfEditDocument(headers = headers, blocks = blocks, footers = footers)
    }

    // MARK: - Header / Footer

    private fun parseHeaderFooterContainer(xml: String, container: String, item: String, plainText: String): List<UdfEditHeaderFooter> {
        val body = extractTaggedBody(xml, openTag = "<$container", closeTag = "</$container>") ?: return emptyList()

        val items = mutableListOf<UdfEditHeaderFooter>()
        var search = 0
        val openItem = "<$item"

        while (true) {
            val start = body.indexOf(openItem, search)
            if (start == -1) break
            val end = body.indexOf("</$item>", start)
            if (end == -1) break
            val itemXml = body.substring(start, end + "</$item>".length)
            val type = extractStringAttr(itemXml, "type") ?: "default"

            val paragraphs = mutableListOf<UdfEditParagraph>()
            var pSearch = 0
            while (true) {
                val pStart = itemXml.indexOf("<paragraph ", pSearch)
                if (pStart == -1) break
                val pEnd = itemXml.indexOf("</paragraph>", pStart)
                if (pEnd == -1) break
                paragraphs.add(parseParagraph(itemXml.substring(pStart, pEnd + "</paragraph>".length), plainText))
                pSearch = pEnd + "</paragraph>".length
            }

            if (paragraphs.isEmpty()) {
                val pStart = itemXml.indexOf("<paragraph")
                if (pStart != -1) {
                    val pEnd = itemXml.indexOf("</paragraph>", pStart)
                    if (pEnd != -1) {
                        paragraphs.add(parseParagraph(itemXml.substring(pStart, pEnd + "</paragraph>".length), plainText))
                    }
                }
            }

            if (paragraphs.isEmpty()) {
                paragraphs.add(UdfEditParagraph(runs = listOf(UdfEditRun(text = ""))))
            }

            items.add(UdfEditHeaderFooter(type = type, paragraphs = paragraphs))
            search = end + "</$item>".length
        }

        return items
    }

    // MARK: - Paragraph

    private fun parseParagraph(xml: String, plainText: String): UdfEditParagraph =
        UdfEditParagraph(
            alignment = extractIntAttr(xml, "Alignment") ?: 3,
            spaceAbove = extractFloatAttr(xml, "SpaceAbove") ?: 1f,
            spaceBelow = extractFloatAttr(xml, "SpaceBelow") ?: 1f,
            leftIndent = extractFloatAttr(xml, "LeftIndent") ?: 0f,
            rightIndent = extractFloatAttr(xml, "RightIndent") ?: 0f,
            firstLineIndent = extractFloatAttr(xml, "FirstLineIndent") ?: 0f,
            hangingIndent = extractFloatAttr(xml, "Hanging") ?: 0f,
            lineSpacing = extractFloatAttr(xml, "LineSpacing") ?: 0f,
            tabStops = parseTabStops(xml),
            runs = parseRuns(xml, plainText),
        )

    private data class TagRun(
        val kind: UdfRunKind,
        val startOffset: Int,
        val length: Int,
        val bold: Boolean,
        val italic: Boolean,
        val underline: Boolean,
        val fontSize: Float?,
        val fontFamily: String?,
        val foreground: Int?,
        val background: Int?,
    )

    private fun parseRuns(xml: String, plainText: String): List<UdfEditRun> {
        val tagged = mutableListOf<TagRun>()
        val patterns = listOf("<content " to "content", "<field " to "field", "<space " to "space")

        for ((pattern, kindTag) in patterns) {
            var search = 0
            while (true) {
                val tagStart = xml.indexOf(pattern, search)
                if (tagStart == -1) break
                val tagEnd = xml.indexOf("/>", tagStart)
                if (tagEnd == -1) break
                val tag = xml.substring(tagStart, tagEnd + 2)

                val start = extractIntAttr(tag, "startOffset")
                val length = extractIntAttr(tag, "length")
                if (start == null || length == null) {
                    search = tagEnd + 2
                    continue
                }

                val kind = when (kindTag) {
                    "field" -> {
                        val name = extractStringAttr(tag, "fieldName")
                            ?: extractStringAttr(tag, "name")
                            ?: extractStringAttr(tag, "FieldName")
                            ?: "alan"
                        UdfRunKind.Field(name)
                    }
                    "space" -> UdfRunKind.Space
                    else -> UdfRunKind.Content
                }

                tagged.add(
                    TagRun(
                        kind = kind,
                        startOffset = start,
                        length = length,
                        bold = tag.contains("bold=\"true\"", ignoreCase = true),
                        italic = tag.contains("italic=\"true\"", ignoreCase = true),
                        underline = tag.contains("underline=\"true\"", ignoreCase = true),
                        fontSize = extractFloatAttr(tag, "size") ?: extractFloatAttr(tag, "Size"),
                        fontFamily = extractStringAttr(tag, "family") ?: extractStringAttr(tag, "Family"),
                        foreground = UdfColorCodec.parse(extractStringAttr(tag, "foreground"))
                            ?: UdfColorCodec.parse(extractStringAttr(tag, "Foreground")),
                        background = UdfColorCodec.parse(extractStringAttr(tag, "background"))
                            ?: UdfColorCodec.parse(extractStringAttr(tag, "Background")),
                    ),
                )
                search = tagEnd + 2
            }
        }

        tagged.sortBy { it.startOffset }

        return tagged.map { tag ->
            UdfEditRun(
                kind = tag.kind,
                text = sliceText(plainText, tag.startOffset, tag.length),
                isBold = tag.bold,
                isItalic = tag.italic,
                isUnderline = tag.underline,
                fontSize = tag.fontSize ?: 12f,
                fontFamily = tag.fontFamily ?: "Times New Roman",
                foregroundARGB = tag.foreground,
                backgroundARGB = tag.background,
            )
        }
    }

    // MARK: - Table

    private fun parseTable(xml: String, plainText: String): UdfEditTable {
        val columnCount = extractIntAttr(xml, "columnCount") ?: 2
        val spansString = extractStringAttr(xml, "columnSpans") ?: ""
        val columnSpans = spansString.split(",").mapNotNull { it.trim().toIntOrNull() }
        val border = extractStringAttr(xml, "border") ?: "borderCell"

        val rows = mutableListOf<UdfEditTableRow>()
        var search = 0
        while (true) {
            val rowStart = xml.indexOf("<row", search)
            if (rowStart == -1) break
            val rowEnd = xml.indexOf("</row>", rowStart)
            if (rowEnd == -1) break
            rows.add(parseTableRow(xml.substring(rowStart, rowEnd + "</row>".length), plainText))
            search = rowEnd + "</row>".length
        }

        return UdfEditTable(
            columnCount = maxOf(columnCount, 1),
            columnSpans = columnSpans,
            border = border,
            rows = rows,
        )
    }

    private fun parseTableRow(xml: String, plainText: String): UdfEditTableRow {
        val rowType = extractStringAttr(xml, "rowType") ?: "dataRow"
        val cells = mutableListOf<UdfEditTableCell>()
        var search = 0
        while (true) {
            val cellStart = xml.indexOf("<cell", search)
            if (cellStart == -1) break
            val cellEnd = xml.indexOf("</cell>", cellStart)
            if (cellEnd == -1) break
            cells.add(parseTableCell(xml.substring(cellStart, cellEnd + "</cell>".length), plainText))
            search = cellEnd + "</cell>".length
        }
        return UdfEditTableRow(rowType = rowType, cells = cells)
    }

    private fun parseTableCell(xml: String, plainText: String): UdfEditTableCell {
        val colspan = extractIntAttr(xml, "colspan") ?: 1
        val rowspan = extractIntAttr(xml, "rowspan") ?: 1
        val fill = UdfColorCodec.parse(extractStringAttr(xml, "fillColor"))
            ?: UdfColorCodec.parse(extractStringAttr(xml, "background"))

        val paragraphs = mutableListOf<UdfEditParagraph>()
        var search = 0
        while (true) {
            val pStart = xml.indexOf("<paragraph ", search)
            if (pStart == -1) break
            val pEnd = xml.indexOf("</paragraph>", pStart)
            if (pEnd == -1) break
            paragraphs.add(parseParagraph(xml.substring(pStart, pEnd + "</paragraph>".length), plainText))
            search = pEnd + "</paragraph>".length
        }

        if (paragraphs.isEmpty()) {
            val pStart = xml.indexOf("<paragraph")
            if (pStart != -1) {
                val pEnd = xml.indexOf("</paragraph>", pStart)
                if (pEnd != -1) {
                    paragraphs.add(parseParagraph(xml.substring(pStart, pEnd + "</paragraph>".length), plainText))
                }
            }
        }

        if (paragraphs.isEmpty()) {
            paragraphs.add(UdfEditParagraph(runs = listOf(UdfEditRun(text = ""))))
        }

        return UdfEditTableCell(
            colspan = maxOf(colspan, 1),
            rowspan = maxOf(rowspan, 1),
            fillColorARGB = fill,
            paragraphs = paragraphs,
        )
    }

    // MARK: - Helpers

    private fun fallbackDocument(text: String): UdfEditDocument {
        val paragraphs = text.split("\n").map { line -> UdfEditParagraph(alignment = 3, runs = listOf(UdfEditRun(text = line))) }
        return UdfEditDocument(blocks = paragraphs.map { UdfEditBlock.ParagraphBlock(it) })
    }

    private fun extractTaggedBody(xml: String, openTag: String, closeTag: String): String? {
        val start = xml.indexOf(openTag)
        if (start == -1) return null
        val openEnd = xml.indexOf(">", start)
        if (openEnd == -1) return null
        val close = xml.indexOf(closeTag, openEnd + 1)
        if (close == -1) return null
        return xml.substring(openEnd + 1, close)
    }

    private fun extractElementsBody(xml: String): String? {
        val start = xml.indexOf("<elements")
        if (start == -1) return null
        val openEnd = xml.indexOf(">", start)
        if (openEnd == -1) return null
        val close = xml.indexOf("</elements>", openEnd + 1)
        if (close == -1) return null
        return xml.substring(openEnd + 1, close)
    }

    private fun sliceText(text: String, start: Int, length: Int): String {
        if (start < 0 || length < 0 || start >= text.length) return ""
        val end = minOf(start + length, text.length)
        if (start >= end) return ""
        return text.substring(start, end)
    }

    private fun parseTabStops(xml: String): List<Float> {
        val tabStr = extractStringAttr(xml, "TabSet") ?: return emptyList()
        val parts = tabStr.split(":")
        val pos = parts.firstOrNull()?.toDoubleOrNull()?.toFloat()
        return if (pos != null && pos > 0) listOf(pos) else emptyList()
    }

    /** Swift'in `String.capitalized`'ının tek-kelime (boşluksuz) davranışı: ilk harf büyük, gerisi küçük. */
    private fun swiftCapitalized(value: String): String =
        if (value.isEmpty()) value else value[0].uppercaseChar() + value.substring(1).lowercase()

    private fun attrKeyVariants(name: String): List<String> = listOf(name, name.lowercase(), swiftCapitalized(name))

    private fun extractIntAttr(xml: String, name: String): Int? {
        for (key in attrKeyVariants(name)) {
            val idx = xml.indexOf("$key=\"")
            if (idx == -1) continue
            val start = idx + key.length + 2
            val end = xml.indexOf("\"", start)
            if (end == -1) continue
            xml.substring(start, end).toIntOrNull()?.let { return it }
        }
        return null
    }

    private fun extractFloatAttr(xml: String, name: String): Float? {
        for (key in attrKeyVariants(name)) {
            val idx = xml.indexOf("$key=\"")
            if (idx == -1) continue
            val start = idx + key.length + 2
            val end = xml.indexOf("\"", start)
            if (end == -1) continue
            xml.substring(start, end).toDoubleOrNull()?.let { return it.toFloat() }
        }
        return null
    }

    private fun extractStringAttr(xml: String, name: String): String? {
        for (key in attrKeyVariants(name)) {
            val idx = xml.indexOf("$key=\"")
            if (idx == -1) continue
            val start = idx + key.length + 2
            val end = xml.indexOf("\"", start)
            if (end == -1) continue
            val value = xml.substring(start, end)
            if (value.isNotEmpty()) return value
        }
        return null
    }
}
